--- /sbin/setup-disk.orig
+++ /sbin/setup-disk
@@ -654,6 +654,10 @@
 	# generate mkinitfs.conf
 	mkdir -p "$mnt"/etc/mkinitfs/features.d
 	echo "features=\"$initfs_features\"" > "$mnt"/etc/mkinitfs/mkinitfs.conf
+	if [ -n "${use_crypt}" ]; then
+		echo "static=\"${CRYPT_KEYFILE}\"" >> "$mnt"/etc/mkinitfs/mkinitfs.conf
+	fi
+
 	if [ -n "$raidmod" ]; then
 		echo "/sbin/mdadm" > "$mnt"/etc/mkinitfs/features.d/raid.files
 		echo "/etc/mdadm.conf" >> "$mnt"/etc/mkinitfs/features.d/raid.files
@@ -1214,6 +1218,9 @@
 	mkfs.$ROOTFS $MKFS_OPTS_ROOT $mkfs_args "$root_dev"
 	mkdir -p "$SYSROOT"
 	mount -t $ROOTFS $root_dev "$SYSROOT" || return 1
+	# if [ -n "${USE_CRYPT}" ]; then
+	# 	cryptsetup luksAddKey --key-file "${CRYPT_KEYFILE}" "${root_dev}" "${CRYPT_KEYFILE}"
+	# fi
 	if [ -n "$boot_dev" ]; then
 		# use separate esp partirion with UEFI but without crypt
 		if [ -n "$USE_EFI" ] && [ -z "$USE_CRYPT" ]; then
@@ -1286,26 +1293,16 @@
 setup_crypt() {
 	local dev="$1" local dmname="$2"
 	mkdir -p /run/cryptsetup
-	while true; do
-		echo "Preparing partition for encryption." >&2
-		echo "You will be prompted for your password at boot." >&2
-		echo "If you forget your password, your data will be lost." >&2
-		cryptsetup luksFormat --batch-mode --verify-passphrase \
-			--type luks2 "$dev" >&2
+
 		# Error codes are: 1 wrong parameters, 2 no permission (bad
 		# passphrase), 3 out of memory, 4 wrong device specified, 5 device
 		# already exists or device is busy.
 		# https://man7.org/linux/man-pages/man8/cryptsetup.8.html#RETURN_CODES
-		case $? in
-			2) continue;;
-			0) ;;
-			*) return 1;;
-		esac
-		echo "Enter password again to unlock disk for installation." >&2
-		cryptsetup open "$dev" "$dmname" >&2 \
-			&& break
-		echo "" >&2
-	done
+	cryptsetup luksFormat --key-file "${CRYPT_KEYFILE}" --batch-mode --type luks2 "$dev" >&2
+
+	cryptsetup open --key-file "${CRYPT_KEYFILE}" "$dev" "$dmname" >&2
+	cryptsetup luksAddKey "/dev/mapper/${dmname}" "${CRYPT_KEYFILE}" >&2
+
 	echo "/dev/mapper/$dmname"
 	return 0
 }
@@ -1592,6 +1589,12 @@
 	arm*|aarch64) : ${BOOTLOADER:=u-boot};;
 esac
 
+if [ -n "${USE_CRYPT}" ]; then
+	CRYPT_KEYFILE="/root/keyfile.bin"
+	dd bs=1024 count=4 if=/dev/urandom of="${CRYPT_KEYFILE}"
+	chmod 0400 "${CRYPT_KEYFILE}"
+fi
+
 if [ -d "$1" ]; then
 	# install to given mounted root
 	apk add --quiet $(select_bootloader_pkg)
@@ -1726,4 +1729,3 @@
 none) exit 0;;
 *) die "Not a valid install mode: $DISK_MODE" ;;
 esac
-
