--- /sbin/setup-disk.orig
+++ /sbin/setup-disk
@@ -374,7 +374,7 @@
 	local exlinux_raidopt=
 
 	sed -e "s:^root=.*:root=$root:" \
-		-e "s:^default_kernel_opts=.*:default_kernel_opts=\"$kernel_opts\":" \
+		-e "s#^default_kernel_opts=.*#default_kernel_opts=\"$kernel_opts\"#" \
 		-e "s:^modules=.*:modules=$modules:" \
 		/etc/update-extlinux.conf > "$mnt"/etc/update-extlinux.conf
 
@@ -645,7 +645,7 @@
 
 	if crypt_required <"$mnt"/etc/fstab; then
 		use_crypt=1
-		initfs_features="$initfs_features cryptsetup keymap"
+		initfs_features="$initfs_features cryptsetup cryptkey keymap"
 		if is_nvme_dev "$ROOT"/sys/block/dm-*/slaves/*; then
 			initfs_features="$initfs_features nvme"
 		fi
@@ -654,6 +654,10 @@
 	# generate mkinitfs.conf
 	mkdir -p "$mnt"/etc/mkinitfs/features.d
 	echo "features=\"$initfs_features\"" > "$mnt"/etc/mkinitfs/mkinitfs.conf
+	if [ -n "${use_crypt}" ]; then
+		echo "${CRYPT_KEYFILE}" > "$mnt"/etc/mkinitfs/features.d/cryptkey.files
+	fi
+
 	if [ -n "$raidmod" ]; then
 		echo "/sbin/mdadm" > "$mnt"/etc/mkinitfs/features.d/raid.files
 		echo "/etc/mdadm.conf" >> "$mnt"/etc/mkinitfs/features.d/raid.files
@@ -677,7 +681,7 @@
 		if cryptsetup status "$cryptroot" 2>&1 >/dev/null; then
 			cryptroot=$(cryptsetup status "$cryptroot" | awk '/device:/ { print $2 }')
 			cryptroot=$(uuid_or_device $cryptroot)
-			kernel_opts="cryptroot=$cryptroot cryptdm=root $kernel_opts"
+			kernel_opts="cryptroot=$cryptroot cryptkey=rootfs:${CRYPT_KEYFILE} cryptdm=root $kernel_opts"
 			root=$([ -n "$pvs" ] && echo "$rootdev" || echo "/dev/mapper/root")
 		fi
 
@@ -1286,26 +1290,15 @@
 setup_crypt() {
 	local dev="$1" local dmname="$2"
 	mkdir -p /run/cryptsetup
-	while true; do
-		echo "Preparing partition for encryption." >&2
-		echo "You will be prompted for your password at boot." >&2
-		echo "If you forget your password, your data will be lost." >&2
-		cryptsetup luksFormat --batch-mode --verify-passphrase \
-			--type luks2 "$dev" >&2
+
 		# Error codes are: 1 wrong parameters, 2 no permission (bad
 		# passphrase), 3 out of memory, 4 wrong device specified, 5 device
 		# already exists or device is busy.
 		# https://man7.org/linux/man-pages/man8/cryptsetup.8.html#RETURN_CODES
-		case $? in
-			2) continue;;
-			0) ;;
-			*) return 1;;
-		esac
-		echo "Enter password again to unlock disk for installation." >&2
-		cryptsetup open "$dev" "$dmname" >&2 \
-			&& break
-		echo "" >&2
-	done
+	cryptsetup luksFormat --key-file "${CRYPT_KEYFILE}" --batch-mode --type luks2 "$dev" >&2
+	cryptsetup luksAddKey --key-file "${CRYPT_KEYFILE}" "$dev" "${CRYPT_KEYFILE}" >&2
+	cryptsetup open --key-file "${CRYPT_KEYFILE}" "$dev" "$dmname" >&2
+
 	echo "/dev/mapper/$dmname"
 	return 0
 }
@@ -1592,6 +1585,12 @@
 	arm*|aarch64) : ${BOOTLOADER:=u-boot};;
 esac
 
+if [ -n "${USE_CRYPT}" ]; then
+	CRYPT_KEYFILE="/root/keyfile.bin"
+	dd bs=1024 count=4 if=/dev/urandom of="${CRYPT_KEYFILE}"
+	chmod 0400 "${CRYPT_KEYFILE}"
+fi
+
 if [ -d "$1" ]; then
 	# install to given mounted root
 	apk add --quiet $(select_bootloader_pkg)
@@ -1726,4 +1725,3 @@
 none) exit 0;;
 *) die "Not a valid install mode: $DISK_MODE" ;;
 esac
-
