--- /sbin/setup-user.orig
+++ /sbin/setup-user
@@ -17,6 +17,7 @@
 		 -g  Comma or space separated list of groups to add user to
 		 -k  ssh key or URL to ssh key (eg. https://gitlab.alpinelinux.org/user.keys)
 		     or 'none' for no key
+		 -s  Set login shell for user
 		 -u  Unlock the user automatically (eg. creating the user non-interactively
 		     with an ssh key for login)
 
@@ -25,20 +26,21 @@
 	exit $1
 }
 
-while getopts "af:g:hk:u" opt; do
+while getopts "af:g:hk:s:u" opt; do
 	case $opt in
 		a) admin=1;;
 		h) usage 0;;
 		f) fullnameopt="$OPTARG";;
 		g) groups="$OPTARG";;
 		k) keysopt="$OPTARG";;
+		s) loginshell="$OPTARG";;
 		u) forceunlock=1;;
 		'?') usage "1" >&2;;
 	esac
 done
 shift $(($OPTIND - 1))
 
-if [ -z "$admin$fullnameopt$groups$keysopt$forceunlock" ] && [ "$1" = "none" ]; then
+if [ -z "$admin$fullnameopt$groups$keysopt$loginshell$forceunlock" ] && [ "$1" = "none" ]; then
 	exit 0
 fi
 
@@ -76,11 +78,15 @@
 		lastfullname="$resp"
 	fi
 
-	if [ -n "$fullname" ]; then
-		adduser -g "$fullname" $nopassword "$username" && break
-	else
-		adduser $nopassword "$username" && break
+	if adduser ${fullname:+-g "$fullname"} ${loginshell:+-s "$loginshell"} $nopassword "$username"; then
+		if [ -n "$nopassword" ]; then
+			# `adduser -D` creates a locked account, which prevents SSH login
+			# This enables ssh login but disallows local password for user
+			# https://unix.stackexchange.com/a/193131
+			sed -i -E "s/^($username:)\!:/\1\*:/" /etc/shadow
 	fi
+		break
+	fi
 	if ! [ -n "$interactive" ]; then
 		exit 1
 	fi
@@ -109,12 +115,12 @@
 				break
 				;;
 			https://*|http://*)
-				sshkeys=$(wget -q -O- $resp | grep ^ssh-)
+				sshkeys=$(wget -q -O- $resp | grep -E "^(ssh|ecdsa|sk-ecdsa)-")
 				;;
 			*)	sshkeys="$resp"
 				;;
 		esac
-		if echo "$sshkeys" | grep -q ^ssh-; then
+		if echo "$sshkeys" | grep -q -E "^(ssh|ecdsa|sk-ecdsa)-"; then
 			break
 		fi
 		echo "Did not find any key in '$resp'"
@@ -122,13 +128,13 @@
 else
 	case "$keysopt" in
 		https://*|http://*)
-			sshkeys=$(wget -q -O- "$keysopt" | grep ^ssh-);;
+			sshkeys=$(wget -q -O- "$keysopt" | grep -E "^(ssh|ecdsa|sk-ecdsa)-");;
 		none)
 			sshkeys="" ;;
 		*)
 			sshkeys="$keysopt";;
 	esac
-	if [ -n "$sshkeys" ] && ! echo "$sshkeys" | grep -q ^ssh-; then
+	if [ -n "$sshkeys" ] && ! echo "$sshkeys" | grep -q -E "^(ssh|ecdsa|sk-ecdsa)-"; then
 		echo "Could not find any keys in '$resp'" >&2
 		exit 1
 	fi
@@ -153,7 +159,7 @@
 if [ -n "$admin" ]; then
 	apk add doas
 	mkdir -p "$ROOT"/etc/doas.d
-	echo "permit persist :wheel" >> "$ROOT"/etc/doas.d/doas.conf
+	echo "permit nopass :wheel" >> "$ROOT"/etc/doas.d/wheel.conf
 	$MOCK addgroup "$username" "wheel" || exit
 fi
 
